fix(zone.js): zone-node only patch Promise.prototype.then

JiaLiPassion opened the pull request:
Close https://github.com/angular/angular/issues/47872

zone-node only patches `Promise.prototype.then` instead of patch
`Promise` itself. So the new NodeJS `SafePromise` will not complain
the Promise.prototype.then called on incompatible receiver.

We should also do this change on browser zone.js patch, but it will
be a big breaking change, because Promise.prototype.then will not work
with `fakeAsync` any longer.

alan-agius4 commented on tools/testing/node_tests.init.ts:
```diff
@@ -15,7 +15,7 @@ import './zone_base_setup';
 // Init TestBed
 import {TestBed} from '@angular/core/testing';
-import {ServerTestingModule, platformServerTesting} from '@angular/platform-server/testing/src/server';
+import {ServerTestingModule, platformServerTesting,} from '@angular/platform-server/testing/src/server';
```
Can you revert this please?

JiaLiPassion replied:
Sure, just reverted.

alan-agius4 reviewed (approved):
LGTM, confirmed that this solved the issue.

JiaLiPassion committed:
fix(zone.js): zone-node only patch Promise.prototype.then

Close #47872

zone-node only patches `Promise.prototype.then` instead of patch
`Promise` itself. So the new NodeJS `SafePromise` will not complain
the Promise.prototype.then called on incompatible receiver.

We should also do this change on browser zone.js patch, but it will
be a big breaking change, because Promise.prototype.then will not work
with `fakeAsync` any longer.

josephperrott reviewed (approved):
LGTM

AndrewKushnir commented on packages/zone.js/lib/node/promise.ts:
```diff
@@ -0,0 +1,81 @@
+ * found in the LICENSE file at https://angular.io/license
+ */
+Zone.__load_patch('ZoneAwarePromise', (global: any, Zone: ZoneType, api: _ZonePrivate) => {
+  const __symbol__ = api.symbol;
```
It'd be great to add a comment into the code here to explain why we have a different implementation for Promise handling and which NodeJS versions are affected.

JiaLiPassion replied:
Thank you for the comment, I have added the comment to explain why this new implementation is needed.

AndrewKushnir commented on packages/zone.js/lib/node/promise.ts:
```diff
@@ -0,0 +1,81 @@
+      }
+      return (val: any) => {
+        const task = zone.scheduleMicroTask(source, () => {
+          return typeof resolveFunc === 'function' ? resolveFunc(val) :
+              isReject                             ? Promise.reject(val) :
+                                                     val;
```
Minor suggestion for readability:
```suggestion
          return typeof resolveFunc === 'function'
              ? resolveFunc(val)
              : (isReject ? Promise.reject(val) : val);
```

JiaLiPassion replied:
Got it, fixed, thank you.

AndrewKushnir reviewed (approved):
@JiaLiPassion thanks for the fix 👍 

I've left a couple minor comments and I also have a question (just to double check): would that version work correctly with the NodeJS version range that we have for Angular apps (starting from v15.2)?

(we can also update the `feat(zone.js)` -> `fix(zone.js)`, so that we can land this change sooner than v16)

JiaLiPassion commented:
@AndrewKushnir, thank you for the review, I have updated the code and also the git comment for the commit.

AndrewKushnir reviewed (approved):
Thanks for addressing the feedback @JiaLiPassion 👍

AndrewKushnir commented:
[Presubmit](http://test/OCL:511528943:BASE:511528984:1677089722825:f461f689).

AndrewKushnir commented:
**Caretaker note**: [this CL](https://critique.corp.google.com/cl/511528176) contains an updated patch for g3.

AndrewKushnir commented:
This PR was merged into the repository by commit d1ac3aa14e5d3c5415937199a6fb63437ddee0b8.
